core/lib.rs

```diff
@@ -6,71 +6,103 @@ extern crate libc;
 mod js_errors;
 mod libdeno;
 mod shared;
+mod shared_simple;
 
 pub use crate::js_errors::*;
 pub use crate::libdeno::deno_buf;
 pub use crate::shared::*;
+pub use crate::shared_simple::*;
 use futures::Async;
 use futures::Future;
 use futures::Poll;
 use libc::c_void;
-use std::collections::HashMap;
 use std::ffi::CStr;
 use std::ffi::CString;
 use std::sync::{Once, ONCE_INIT};
 
-pub struct Isolate {
+pub struct Isolate<R = SharedSimpleRecord, S = SharedSimple> {
   libdeno_isolate: *const libdeno::isolate,
-  pending_ops: HashMap<i32, PendingOp>, // promise_id -> op
+  pending_ops: Vec<PendingOp<R>>,
   polled_recently: bool,
-  recv_cb: RecvCallback,
+  recv_cb: RecvCallback<R, S>,
 
-  pub shared: Shared,
+  pub shared: S,
   pub test_send_counter: u32, // TODO only used for testing- REMOVE.
 }
 
-pub type RecvCallback = fn(isolate: &mut Isolate, zero_copy_buf: deno_buf);
+pub type RecvCallback<R, S> =
+  fn(isolate: &mut Isolate<R, S>, zero_copy_buf: deno_buf);
 
-pub const NUM_RECORDS: usize = 100;
+/// Buf represents a byte array returned from a "Op".
+/// The message might be empty (which will be translated into a null object on
+/// the javascript side) or it is a heap allocated opaque sequence of bytes.
+/// Usually a flatbuffer message.
+pub type Buf = Box<[u8]>;
 
-// TODO rename to AsyncResult
-pub struct AsyncResult {
-  pub result: i32,
-}
-
-pub type Op = dyn Future<Item = AsyncResult, Error = std::io::Error> + Send;
+/// JS promises in Deno map onto a specific Future
+/// which yields a Buf object. Ops should never error.
+/// Errors need to be encoded into the Buf and sent back to JS.
+pub type Op<R> = dyn Future<Item = R, Error = ()> + Send;
 
-struct PendingOp {
-  op: Box<Op>,
+struct PendingOp<R> {
+  op: Box<Op<R>>,
   polled_recently: bool,
   zero_copy_id: usize, // non-zero if associated zero-copy buffer.
+  result: Option<R>,
+}
+
+impl<R> PendingOp<R> {
+  /// Calls poll() on a PendingOp. Returns true if it's complete.
+  fn is_complete(self: &mut PendingOp<R>) -> bool {
+    // TODO the name of this method isn't great, because it doesn't indicate
+    // that the op is being polled.
+
+    // Do not call poll on futures we've already polled this turn.
```

Comment:
How come we are polling the same future twice on the same turn? 
(I'm worried that that doesn't scale well if there are many outstanding ops)

Reply:
Ops aren’t polled more than once. 

Imagine the situation where an accept is followed by a read. Once the accept has completed and before the read has been added, mio has not been notified to wake up. Failing to poll the newly added read op will result in incorrect behavior.